conn.go

```diff
@@ -166,9 +166,9 @@ var TimeoutLimit int64 = 0
 // queries, but users are usually advised to use a more reliable, higher
 // level API.
 type Conn struct {
-	conn net.Conn
-	r    *bufio.Reader
-	w    contextWriter
+	connWrapper *ConnWrapper
+	r           *bufio.Reader
+	w           contextWriter
```

Comment:
Let's start with proper abstraction.
Let's have a one interface, call it `NetworkConnection`:
```
type NetworkConnection interface {
	SetWriteDeadline(time.Time) error
	WriteContext(ctx context.Context, p []byte) (n int, err error)
	io.Reader
	io.Closer
	RemoteAddr() net.Addr
}
```

It is going to be used as following:
```
type Conn struct {
	nc NetworkConnection
}
```

Let's have factory on the `ClusterConfig`
```
type NetworkConnectionFactory interface {
	Wrap(conn net.Conn) NetworkConnection
}
```

Now, `DefaultNetworkConnectionFactory` would be wrapping writer with `deadlineContextWriter` and reader with `bufio.NewReader`